File: src/players/minimax.js
import player from './player';
import winner from '../algorithms/winner';

const RED = 1;
const BLUE = 2;

const INFINITY = 1000000;

class minimax extends player {
  constructor(n, heuristic) {
    super(n);
    this.heuristic = heuristic;
  }

  abMinimax(board, whoami) {
    const oponent = whoami === RED ? BLUE : RED;
  
    const rec = (depth, max, alpha, beta) => {
      const turn = max ? whoami : oponent;
      const other = turn === RED ? BLUE : RED;
      if (depth === 0 || winner(board)) {
        return [this.heuristic(board, max ? turn : other)];
      }

      if (max) {
        let value = -INFINITY;
        let move = null;

        for (let x = 0; x < this.n; x++) {
          for (let y = 0; y < this.n; y++) {
            if (!board[x][y]) {
              board[x][y] = turn;
              const [v] = rec(depth - 1, false);
              if (v > value) {
                value = v;
                move = [x, y];
              }
              alpha = Math.max(alpha, value);
              board[x][y] = 0;
              if (alpha >= beta) {
                return [value, move];
              }
            }
          }
        }

        return [value, move];
      }

      let value = INFINITY;
      let move = null;
      for (let x = 0; x < this.n; x++) {
        for (let y = 0; y < this.n; y++) {
          if (!board[x][y]) {
            board[x][y] = turn;
            const [v] = rec(depth - 1, true);
            if (v < value) {
              value = v;
              move = [x, y];
            }
            beta = Math.min(beta, value);
            board[x][y] = 0;
            if (alpha >= beta) {
              return [value, move];
            }
          }
        }
      }

      return [value, move];
    }

    return rec(2, true, -INFINITY, INFINITY);
  }

  play(board, whoami, _last, callback) {
    const [v, m] = this.abMinimax(board, whoami);
    console.log(`v = ${v}`);
    callback(m[0], m[1]);
  }
}

export default minimax;

File: ai/minimax.go
package ai

func (p *Player) minimax(depth int, player PlayerID, board *Board) (*Move, int) {
	var oponent = Max
	best := infinity + 1
	if player == Max {
		oponent = Min
		best = -infinity - 1
	}

	// Check if board has a winner.
	winner := Winner(board)
	if winner == Max {
		return nil, infinity
	} else if winner == Min {
		return nil, -infinity
	}

	// No more depth to recurse; return heuristic.
	if depth == 0 {
		v := p.Heuristic(board)
		return nil, v
	}

	var mv *Move
	for i := 0; i < board.Size; i++ {
		for j := 0; j < board.Size; j++ {
			if board.Matrix[i][j] == NoOne {
				board.Matrix[i][j] = player
				_, v := p.minimax(depth-1, oponent, board)
				board.Matrix[i][j] = NoOne

				if player == Max && v > best {
					best = v
					mv = &Move{i, j}
				} else if player == Min && v < best {
					best = v
					mv = &Move{i, j}
				}
			}
		}
	}

	return mv, best
}

// Minimax runs minimax in the given board, until the given depth, and
// returns a move and its expected outcome.
func (p *Player) Minimax(board Board, depth int) (*Move, int) {
	return p.minimax(depth, p.ID, &board)
}
